### PR TITLE
Add overview document for clustering

### PR DESCRIPTION
**Pull request by arovir01**
This PR adds the overview document for clustering.

This is still to be viewed as work in progress until the tutorial document will have been finalized.

### CODE REVIEW

File: tensorflow_model_optimization/g3doc/guide/clustering/index.md [hunk 1]
```diff
@@ -0,0 +1,103 @@
+# Weight Clustering
```
Comment (alanchiao) [inline tensorflow_model_optimization/g3doc/guide/clustering/index.md:1]: I'd update this based on https://www.tensorflow.org/model_optimization/guide/quantization/training
as a reference point.

Comment (alanchiao) [inline tensorflow_model_optimization/g3doc/guide/clustering/index.md:1]: "Weight clustering" to match TF style

File: tensorflow_model_optimization/g3doc/guide/clustering/index.md [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+This document provides an overview on weight clustering to help you determine how it fits with your use case. To dive right into the code, see the [Clustering with Keras](clustering_with_keras.ipynb) tutorial and the [API docs](../../api_docs/python). For additional details on how to use the Keras API, a deep dive into weight clustering, and documentation on more advanced usage patterns, see the [API usage guide](train_clustered_models.md).
+
+## Overview
+
+Clustering, or weight sharing, is a technique for reducing the number of unique weight values in a model. It first groups the weights of each layer into *N* clusters, then shares the cluster's centroid value for all the weights belonging to the cluster.
```
Comment (alanchiao) [inline tensorflow_model_optimization/g3doc/guide/clustering/index.md:7]: For conciseness: -> "Clustering, or weight sharing, reduces the number of unique weight values in a model."

"Clustering, or weight sharing, reduces the number of unique weight values in a model, leading to benefits for deployment." to mirror the quantization docs

File: tensorflow_model_optimization/g3doc/guide/clustering/index.md [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+
+This technique brings improvements in terms of model compression. It reduces the number of bits used to represent the weights, thus saving memory bandwidth. This can be critical for deploying deep learning models on embedded systems with limited resources.
```
Comment (alanchiao) [inline tensorflow_model_optimization/g3doc/guide/clustering/index.md:9]: Does it reduce the number of bits used to represent the weights?

Thought it'd still be float32.

> ... saving memory bandwidth
We can use a similar phrasing from the proposal RFC to first state what benefits apply to everyone. It should then be clarified that memory bandwidth saving requires framework support, similar to what's mentioned for pruning. https://www.tensorflow.org/model_optimization/guide/pruning

File: tensorflow_model_optimization/g3doc/guide/clustering/index.md [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+### API Compatibility Matrix
+
+Users can apply clustering with the following APIs:
+
+*   Model building: `tf.keras` with only Sequential and Functional models
+*   TensorFlow versions: TF 1.x for versions 1.14+ and 2.x.
+    *   `tf.compat.v1` with a TF 2.X package and `tf.compat.v2` with a TF 1.X
+        package are not supported.
+*   TensorFlow execution mode: both graph and eager
+*   Distributed training: `tf.distribute` with only graph execution
```
Comment (alanchiao) [inline tensorflow_model_optimization/g3doc/guide/clustering/index.md:22]: I think tf.distribute was considered out of scope of the initial release right? 

Please update this matrix once it's finalized in the RFC review

File: tensorflow_model_optimization/g3doc/guide/clustering/index.md [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+3. Compress the .tflite file into a zip
+
+## Examples
+
+In addition to the [Clustering with Keras](clustering_with_keras.ipynb) tutorial, see the following examples:
+
+* Cluster the weights of a CNN model trained on the MNIST handwritten digit classification databaset:
+[code](https://github.com/tensorflow/model-optimization/blob/master/tensorflow_model_optimization/python/examples/clustering/keras/mnist/mnist_cnn.py)
+
+## Tips
```
Comment (alanchiao) [inline tensorflow_model_optimization/g3doc/guide/clustering/index.md:92]: From the discussion, the tips will no longer live here. Instead, it should be in the comprehensive guide in some form.

File: tensorflow_model_optimization/g3doc/guide/clustering/index.md [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+
+1. The current clustering API works only with pre-trained models. Don't forget to train your model before attempting to cluster it.
+2. The centroid initialization technique you opt for has a significant impact on the accuracy of the clustered model. Experiments have shown that linear initialization outperforms density-based and random initialization in most cases.
+
+## References
+
+The weight clustering implementation is based on the technique described in chapter 3, entitled *Trained Quantization and Weight Sharing*, of the conference paper referenced below.
```
Comment (alanchiao) [inline tensorflow_model_optimization/g3doc/guide/clustering/index.md:99]: entitled -> titled

Remove ## References header to match styling.

File: tensorflow_model_optimization/g3doc/guide/clustering/index.md (hunk 5 above)
Comment (arovir01) [reply tensorflow_model_optimization/g3doc/guide/clustering/index.md:93]: As the comprehensive guide already contains these tips, I will simply remove them from here.

File: tensorflow_model_optimization/g3doc/guide/clustering/index.md [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+## Examples
+
+In addition to the [Clustering with Keras](clustering_with_keras.ipynb) tutorial, see the following examples:
+
+* Cluster the weights of a CNN model trained on the MNIST handwritten digit classification databaset:
+[code](https://github.com/tensorflow/model-optimization/blob/master/tensorflow_model_optimization/python/examples/clustering/keras/mnist/mnist_cnn.py)
+
+## References
+
+The weight clustering implementation is based on the technique described in chapter 3, titled *Trained Quantization and Weight Sharing*, of the conference paper referenced below.
```
Comment (alanchiao) [inline tensorflow_model_optimization/g3doc/guide/clustering/index.md:95]: Remove #References for consistency and keep this to the point and in order of what a reader of this library needs (link earlier).

The weight clustering implementation is based on the *Deep Compression: Compressing Deep Neural Networks With Pruning, Trained Quantization and Huffman Coding* [paper](https://arxiv.org/abs/1510.00149). See chapter 3, titled *Trained Quantization and Weight Sharing*.

Would welcome alternatives for these references

File: tensorflow_model_optimization/g3doc/guide/clustering/index.md [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+The models were trained and tested on SpeechCommands v0.02.
+
+NOTE: *Size of compressed .tflite* refers to the size of the zipped .tflite file obtained from the model through the following process:
+1. Serialize the Keras model into .h5 file
+2. Convert the .h5 file into .tflite using `TFLiteConverter.from_keras_model_file()`
+3. Compress the .tflite file into a zip
+
+## Examples
+
+In addition to the [Clustering with Keras](clustering_with_keras.ipynb) tutorial, see the following examples:
```
Comment (alanchiao) [inline tensorflow_model_optimization/g3doc/guide/clustering/index.md:88]: Nit: tutorial -> example 

To emphase it's just one end-to-end flow.

File: tensorflow_model_optimization/g3doc/guide/clustering/index.md [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+    <td>1.5 MB</td>
+    <td>32</td>
+    <td>94.71%</td>
+    <td>0.3 MB</td>
+  </tr>
+</table>
+
+The models were trained and tested on SpeechCommands v0.02.
+
+NOTE: *Size of compressed .tflite* refers to the size of the zipped .tflite file obtained from the model through the following process:
```
Comment (alanchiao) [inline tensorflow_model_optimization/g3doc/guide/clustering/index.md:81]: Edit for conciseness: "file obtained from the model through the following process" -> "file obtained from the following process"

File: tensorflow_model_optimization/g3doc/guide/clustering/index.md [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+
+We have seen up to 5x improvements in model compression with minimal loss of accuracy, as demonstrated by the [results](#results) presented below. The compression gains depend on the model and the accuracy targets in each specific use case. For example, for the MobileNetV2 image classification model, one can choose to reduce all non-depthwise convolutional layers to use just 32 unique weigth values and obtain a float32 tflite model that is approximately 4.8 times more compressible using ZIP Deflate algorithm than the original model. However, that will result in about 3% drop of the top-1 classification accuracy. On the other hand, the same model clustered less agressively, using 256 clusters for two internal layers and 32 clusters for the final convolutional layer, maintains virtually the same accuracy as the original model, yet still yields a respectable 1.8x improvement in compression ratio.
```
Comment (alanchiao) [inline tensorflow_model_optimization/g3doc/guide/clustering/index.md:11]: On "the compression gains depend on the model and the accuracy targets in each specific use case. ... yet still yields a respectable 1.8x improvement in compression ratio."

I'd update the table based on what you did in the RFC to provide these details, and then remove the paragraph. The table gets the message across better.

Comment (alanchiao) [inline tensorflow_model_optimization/g3doc/guide/clustering/index.md:11]: Like in the QAT example, some short summary of the use cases.

"This technique has been experimented across vision and speech tasks."

File: tensorflow_model_optimization/g3doc/guide/clustering/index.md [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+
+Clustering works well with TFLiteConverter, providing an easy path to produce deployment-ready models that can be easily compressed using either an off-the-shelf compression algorithm, similar to the ZIP Deflate we use for demonstration in this document, or a custom method optimized for a special target hardware. When converting the clustered model with TFLiteConverter, the actual number of unique weight values per tensor may increase. This happens for the models with batch normalization layers that are folded into the preceding convolutional layers during the conversion, and also due to different scale factors in the per-channel weight quantization scheme. Both techniques may alter the same weight value differently, depending on the channel it appears in and the associated batch-normalization and quantization parameters. While this side effect may result in a slightly lower compression ratio, the overall benefits of using clustering and post-training conversion and quantization are still tangible, as demonstrated by the examples in this document.
```
Comment (alanchiao) [inline tensorflow_model_optimization/g3doc/guide/clustering/index.md:13]: For a user making a decision, a lot of these details don't help with deciding whether to use weight clustering. 

"TFLiteConverter, providing an easy path to produce deployment-ready models that can be easily compressed using either an off-the-shelf compression algorithm, similar to the ZIP Deflate"

This goes into the "how to use", which users should just reference the tutorials for. 

Something to the point for this whole paragraph is:
"Clustering will provide useful but reduced benefits for convolution and dense layers that precede a batch normalization layer and in combination with per-axis post-training quantization."

They can otherwise see the consequences from the Mobilenet v1 and v2 numbers in the below, that face the batchnorm folding issue. If you can quantify the consequences with a range (Nx - Mx instead of the usual Px - Qx), that'd help also.

The details (e.g. a specific one with the math) may be interesting, but in a "Internal" page for people who want to understand how things work. We haven't created this page yet for other techniques, but will need to once we move the quantization scheme over to the TFMOT docs. 

@Ruomei

File: tensorflow_model_optimization/g3doc/guide/clustering/index.md [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+
+This document provides an overview on weight clustering to help you determine how it fits with your use case. To dive right into the code, see the [weight clustering end-to-end example](clustering_example.ipynb) and the [API docs](../../api_docs/python). For additional details on how to use the Keras API, a deep dive into weight clustering, and documentation on more advanced usage patterns, see the [weight clustering comprehensive guide](clustering_comprehensive_guide.ipynb).
+
+## Overview
+
+Clustering, or weight sharing, reduces the number of unique weight values in a model, leading to benefits for deployment. It first groups the weights of each layer into *N* clusters, then shares the cluster's centroid value for all the weights belonging to the cluster.
+
+This technique brings improvements in terms of model compression. By reducing the number of unique weight values, weight clustering renders the weights suitable for compression via Huffman coding and similar techniques. Future framework support will, therefore, be able to reduce memory size and bandwidth requirements. This can be critical for deploying deep learning models on embedded systems with limited resources.
```
Comment (alanchiao) [inline tensorflow_model_optimization/g3doc/guide/clustering/index.md:9]: 1. Given comment below on "We have seen up to 5x improvements" which will make things much shorter, I think you can fold that content back into this paragraph up here. 

2. "By reducing the number of unique weight values, weight clustering renders the weights suitable for compression via Huffman coding and similar techniques."
I don't think this is needed. The intuition is stated in the first paragraph already. Huffman coding is a part of the "how" and doesn't need to be mentioned in the doc. 
 
E.g. 

This technique brings improvements via model compression. Future framework support can unlock memory footprint improvements. We've seen up to 5x improvements in model compression with minimal loss of accuracy.

File: tensorflow_model_optimization/g3doc/guide/clustering/index.md [hunk 13]
```diff
@@ -0,0 +1,99 @@
+# Weight clustering
```
[the remark this answers is not in the document]
Comment (akarmi) [reply tensorflow_model_optimization/g3doc/guide/clustering/index.md:1]: Good question. It appears that the original paper uses both 'weight sharing' and 'clustering' (on its own), while arguably refering to 'clustering' in a slightly narrower context of initial clustering of weights before the fine tuning. It seems to me the boundary is quite vague but it would make sense to refine the use of these terms.
@arovir01, how about changing the title of this section to 'Clustering', and introduce the interchangeability of both terms in this paragraph, instead of the beginning of the 'Overview' section below? E.g.
"overview on weight clustering to help" -> overview on clustering, or weight sharing, to help"

[the remark this answers is not in the document]
Comment (arovir01) [reply tensorflow_model_optimization/g3doc/guide/clustering/index.md:1]: @akarmi Yes, I can do that.

[the remark this answers is not in the document]
Comment (arovir01) [reply tensorflow_model_optimization/g3doc/guide/clustering/index.md:1]: I believe the "Overview" title should be kept to stay in line with other similar documents, such as https://github.com/tensorflow/model-optimization/blob/master/tensorflow_model_optimization/g3doc/guide/quantization/training.md

File: tensorflow_model_optimization/g3doc/guide/clustering/index.md [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+
+*   Model building: `tf.keras` with only Sequential and Functional models
+*   TensorFlow versions: TF 1.x for versions 1.14+ and 2.x.
+    *   `tf.compat.v1` with a TF 2.X package and `tf.compat.v2` with a TF 1.X
+        package are not supported.
+*   TensorFlow execution mode: both graph and eager
+
+## Results
+
+### Image classification
```
[the remark this answers is not in the document]
Comment (arovir01) [reply tensorflow_model_optimization/g3doc/guide/clustering/index.md:27]: Compiling the latest results requires some further work. I will update them in a later commit.

[the remark this answers is not in the document]
Comment (alanchiao) [reply tensorflow_model_optimization/g3doc/guide/clustering/index.md:30]: Ok

**Comment by alanchiao** [comment]
@lamberta: could you review this PR? This is for a new weight clustering technique that we're moving forward with from [this RFC](https://github.com/tensorflow/model-optimization/blob/master/community/rfcs/techniques/design/20200507-weight-clustering-design.md).

File: tensorflow_model_optimization/g3doc/guide/clustering/index.md [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+NOTE: *Size of compressed .tflite* refers to the size of the zipped .tflite file obtained from the model from the following process:
+1. Serialize the Keras model into .h5 file
+2. Convert the .h5 file into .tflite using `TFLiteConverter.from_keras_model_file()`
+3. Compress the .tflite file into a zip
+
+## Examples
+
+In addition to the [Weight clustering in Keras example](clustering_example.ipynb.ipynb), see the following examples:
+
+* Cluster the weights of a CNN model trained on the MNIST handwritten digit classification databaset:
```
Comment (alanchiao) [inline tensorflow_model_optimization/g3doc/guide/clustering/index.md:93]: Minor typo: databaset -> dataset

**Comment by alanchiao** [comment]
> is the intention to add more docs to a /clustering section?
Yes there is. 

> If this shouldn't get published yet, you can prefix the file name with an underscore like: _index.md
Yes it's not ready for users to view. The idea is to not link it into the left nav via _book.yaml so people cannot really view it on tensorflow.org. If underscore is preferred, we can do that for TFMOT going forward. Only prior to launch, would the _book.yaml changes be made.

**Comment by lamberta** [comment]
> If underscore is preferred ...

Up to you. If only working in GitHub, I'd probably recommend a feature branch or something and then merge everything when ready to publish. Files prefixed with an underscore are skipped when importing---but that also means the file won't get executed (for testing).

It's fine to publish as normal and then add a nav link when ready, as long as you're not leaking anything, etc. (But if that were the case, we wouldn't be chatting out here in GitHub :)

**Comment by alanchiao** [comment]
Ok thanks. Think it's fine to keep the file name as is then. 

It matters less for this file, but for the colabs (written by another person), it's nice to be able to get things automatically tested and reviewed document by document.

**Comment by arovir01** [comment]
> Also need to update the leftnav file: https://github.com/tensorflow/model-optimization/blob/master/tensorflow_model_optimization/g3doc/_book.yaml

Is this something I should do or will one of you guys handle it after merging this?

**Comment by arovir01** [comment]
Please advise if there is anything else that you consider needs changing. Otherwise I will proceed with squashing the commits together in preparation for merging.

**Comment by lamberta** [comment]
@arovir01 for this PR, it sounds like the entry in _book.yaml will be added later by @alanchiao

**Comment by alanchiao** [comment]
Billy, yes that's correct. Or Aron / the others will make it shortly prior to launch.

**Comment by alanchiao** [comment]
Just leaving a note that Billy approved this PR internally.
